hamt/hamt.go

```diff
@@ -458,16 +473,31 @@ func makeAsyncTrieGetLinks(dagService ipld.DAGService, onShardValue func(shard *
 				if err != nil {
 					return nil, err
 				}
-				err = onShardValue(sv)
-				if err != nil {
-					return nil, err
-				}
+				formattedLink := sv.val
+				formattedLink.Name = sv.key
+				emitResult(ctx, linkResults, format.LinkResult{Link: formattedLink, Err: nil})
 			}
 		}
 		return childShards, nil
 	}
 }
 
+func emitResult(ctx context.Context, linkResults chan<- format.LinkResult, r format.LinkResult) {
+	// make sure that context cancel is processed first
+	// the reason is due to the concurrency of EnumerateChildrenAsync
+	// it's possible for EnumLinksAsync to complete and close the linkResults
+	// channel before this code runs
+	select {
```

Comment:
Can these two `select`s be merged?

Reply:
Those `select`s are different atomic operations, checking context cancellation in the first one doesn't guarantee that the context won't be cancelled in the second one (although unlikely, we need to define how sure we want to be of that not happening).

> You specifically want to handle context cancellation first, due to some of the weird parallelism of EnumerateChildrenAsync.

Could you expand on that? If `linkResults` is already closed then the `select` will go to `ctx.Done`. and if it's still open (for whatever reason) what's the harm in sending the result to the channel even if the context might be cancelled?

Reply:
Hey @schomatis thank you for digging in with this cause I should provide a full explanation of how I got here.

So, my understanding is select does not ignore closed channels, and writing to a closed channel causes a panic. I found that was what I was getting at first. Putting the initial statement above solved this issue. Here's my belief about why:

I believe the issue is with `EnumerateChildrenAsyncDepth` - https://github.com/ipfs/go-merkledag/blob/master/merkledag.go#L363 . If you look at the code of this function, there two select blocks, one that happens in the calling goroutine and the other that happens in up to 8 seperate goroutines. All check for ctx.Done(). What I was finding is that the main function was returning to https://github.com/ipfs/go-unixfs/pull/39/files#diff-e4937a94dde5b5b5bd33c42652d0340cR434 which in turn caused the linkResults channel to close. (https://github.com/ipfs/go-unixfs/pull/39/files#diff-e4937a94dde5b5b5bd33c42652d0340cR430) However, I was finding that some of the goroutines in the first block of `EnumerateChildrenAsyncDepth` were still running one more loop: https://github.com/ipfs/go-merkledag/blob/master/merkledag.go#L386 which in turn was calling getLinks which in turn cause our channel to write after close. `EnumerateChildrenAsyncDepth` although it is a parallel function, normally is called synchronously in the sense that the overall routine is not done until all of the children are iterated. I think probably what is ultimately needed is a waitGroup or something at the end of that `EnumerateChildrenAsyncDepth` just to make sure everything is finished before the function returns, so that in this case with a context.Cancel(), the calling function can wait to close it's channels. Anyway, @whyrusleeping does this analysis make any sense?

Maybe I am missing something tho. @magik6k you're welcome to comment at as well if you have thoughts.

Assuming I've got all this right, my inclination is merge and file an issue on go-merkledag and tackle that seperately.

Reply:
Oh and just to be clear this was coming up in practice (the panic) with https://github.com/ipfs/go-ipfs/pull/5611

Reply:
> So, my understanding is select does not ignore closed channels, and writing to a closed channel causes a panic.

Oh, I didn't know that, thanks for the clarification.